cpu/nrf52/radio/nrf802154/nrf802154.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright (C) 2018 Freie Universität Berlin
+ *               2018 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_nrf52_802154
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the radio driver for nRF52 radios
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ * @author      Semjon Kerner <semjon.kerner@fu-berlin.de>
+ * @}
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "luid.h"
+#include "mutex.h"
+
+#include "net/ieee802154.h"
+#include "net/netdev/ieee802154.h"
+#include "nrf802154.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static const netdev_driver_t nrf802154_netdev_driver;
+
+netdev_ieee802154_t nrf802154_dev = {
+    {
+        .driver = &nrf802154_netdev_driver,
+        .event_callback = NULL,
+        .context = NULL,
+#ifdef MODULE_NETSTATS_L2
+        .stats = { 0 }
+#endif
+    },
+#ifdef MODULE_GNRC
+#ifdef MODULE_GNRC_SIXLOWPAN
+    .proto = GNRC_NETTYPE_SIXLOWPAN,
+#else
+    .proto = GNRC_NETTYPE_UNDEF,
+#endif
+#endif
+    .pan = IEEE802154_DEFAULT_PANID,
+    .short_addr = { 0, 0 },
+    .long_addr = { 0, 0, 0, 0, 0, 0, 0, 0 },
+    .chan = IEEE802154_DEFAULT_CHANNEL,
+    .flags = 0
+};
+
+static uint8_t rxbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* len PHR + PSDU + LQI */
+static uint8_t txbuf[IEEE802154_FRAME_LEN_MAX + 3]; /* len PHR + PSDU + LQI */
+
+#define NRF_RX_COMPLETE (0x1)
+#define NRF_TX_COMPLETE (0x2)
+static volatile uint8_t _state;
+static mutex_t txlock;
+
+/**
+ * @brief   Set radio into DISABLED state
+ */
+static void _disable(void)
+{
+    /* set device into DISABLED state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_Disabled) {
+        NRF_RADIO->EVENTS_DISABLED = 0;
+        NRF_RADIO->TASKS_DISABLE = 1;
+        while (!(NRF_RADIO->EVENTS_DISABLED)) {};
+        DEBUG("[nrf802154] Device state: DISABLED\n");
+    }
+}
+
+/**
+ * @brief   Set radio into RXIDLE state
+ */
+static void _enable_rx(void)
+{
+    DEBUG("[nrf802154] Set device state to RXIDLE\n");
+    NRF_RADIO->PACKETPTR = (uint32_t)rxbuf;
+    /* set device into RXIDLE state */
+    if (NRF_RADIO->STATE != RADIO_STATE_STATE_RxIdle) {
+        _disable();
+    }
+    NRF_RADIO->EVENTS_RXREADY = 0;
+    NRF_RADIO->TASKS_RXEN = 1;
+    while (!(NRF_RADIO->EVENTS_RXREADY)) {};
+    DEBUG("[nrf802154] Device state: RXIDLE\n");
+}
+
+/**
+ * @brief   Set radio into TXIDLE state
+ */
+static void _enable_tx(void)
+{
+    DEBUG("[nrf802154] Set device state to TXIDLE\n");
+    NRF_RADIO->PACKETPTR = (uint32_t)txbuf;
```

Comment:
This lightly smells like a race condition here. Is it possible that right after the address is changed here to point to the TX buffer, but while the radio is still in RX mode, a frame is received and accidentally written to the TX buffer instead of the RX buffer?

To be on the safe side, wouldn't it be better to move this register write below the `_disable()` call?
https://github.com/RIOT-OS/RIOT/blob/427111035fcbed0947efe283872d6ca26f2c103d/cpu/nrf52/radio/nrf802154/nrf802154.c#L109-L112